be/src/storage/persistent_index.cpp

```diff
@@ -743,34 +766,80 @@ Status PersistentIndex::load(const PersistentIndexMetaPB& index_meta) {
+    _dump_snapshot = false;
     return Status::NotSupported("TODO");
 }
 
 Status PersistentIndex::commit(PersistentIndexMetaPB* index_meta) {
     // TODO: l0 may be need to flush
-    MutableIndexMetaPB* l0_meta = index_meta->mutable_l0_meta();
-    IndexWalMetaPB* wal_pb = l0_meta->add_wals();
-    _version.to_pb(wal_pb->mutable_version());
+    if (_dump_snapshot) {
+        // if _map size is small enough to dump directly, rewrite snapshot
+        std::string file_name = _get_l0_index_file_name(_path, _version);
+        // be maybe crash after create index file during last commit
+        // so we delete expired index file first to make sure no garbage left
+        Env::Default()->delete_file(file_name);
+        size_t snapshot_size = _dump_bound();
+        phmap::BinaryOutputArchive ar_out(file_name.data());
+        if (!_dump(ar_out)) {
+            std::string err_msg = strings::Substitute("faile to dump snapshot to file $0", file_name);
+            LOG(WARNING) << err_msg;
+            return Status::InternalError(err_msg);
+        }
+        // update PersistentIndexMetaPB
+        MutableIndexMetaPB* l0_meta = index_meta->mutable_l0_meta();
+        l0_meta->clear_wals();
+        IndexSnapshotMetaPB* snapshot = l0_meta->mutable_snapshot();
+        _version.to_pb(snapshot->mutable_version());
+        PagePointerPB* data = snapshot->mutable_data();
+        data->set_offset(0);
+        data->set_size(snapshot_size);
+        _offset += snapshot_size;
+        _page_size = 0;
+    } else {
+        MutableIndexMetaPB* l0_meta = index_meta->mutable_l0_meta();
+        IndexWalMetaPB* wal_pb = l0_meta->add_wals();
+        _version.to_pb(wal_pb->mutable_version());
+
+        PagePointerPB* data = wal_pb->mutable_data();
+        data->set_offset(_offset);
+        data->set_size(_page_size);
 
-    PagePointerPB* data = wal_pb->mutable_data();
-    data->set_offset(_offset);
-    data->set_size(_page_size);
+        _version.to_pb(index_meta->mutable_version());
+        index_meta->set_size(_size);
 
-    _version.to_pb(index_meta->mutable_version());
-    index_meta->set_size(_size);
+        _offset += _page_size;
+        _page_size = 0;
+    }
+    return Status::OK();
+}
+
+Status PersistentIndex::on_commited() {
+    if (_dump_snapshot) {
+        std::string expired_file_path = _index_block->path();
+        std::string index_file_path = _get_l0_index_file_name(_path, _version);
+        fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+        std::unique_ptr<fs::WritableBlock> wblock;
+        fs::CreateBlockOptions wblock_opts({index_file_path});
+        // new index file should be created in commit() phase
+        wblock_opts.mode = Env::MUST_EXIST;
+        RETURN_IF_ERROR(block_mgr->create_block(wblock_opts, &wblock));
+        _index_block = std::move(wblock);
+        VLOG(1) << "delete expired l0 index file: " << expired_file_path;
+        Env::Default()->delete_file(expired_file_path);
```

Comment:
What's the difference between delete_file() and _delete_expired_index_file()

Reply:
> What's the difference between delete_file() and _delete_expired_index_file()

`delete_file()` just delete the old `snapshot_file` when we write a new `snapshot_file` success, but `_delete_expired_index_file()` will delete all mutable index file expect the newest mutable index file. 
IF be crash repeatedly, there may be not only one expired index file, this function will traversing the entire folder and delete all expired mutable index file.